Ensure `wp_editor_set_quality` filter consistently passes output mime type

adamsilverstein opened the pull request:
Fix in place, tests pass.

<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->

Trac ticket: https://core.trac.wordpress.org/ticket/56442

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

adamsilverstein committed:
Add tests for wp_generate_attachment_metadata / wp_editor_set_quality

Tests fail currently, demonstrating issue

adamsilverstein committed:
remove error logging

adamsilverstein committed:
alignment for phpcs

adamsilverstein committed:
Clean up tests, ensure filters in place at beginning

adamsilverstein committed:
set up output format to ensure set_quality uses correct output type

adamsilverstein committed:
move format check into core image class

adamsilverstein committed:
Test doc cleanup

adamsilverstein commented:
Hmm, some other tests fail now because they assert that quality will not be set until save (see `test_set_quality_with_image_conversion`), I'm going to think and discuss more about the best approach here.

adamsilverstein committed:
Set quality more consistently in WP_Image_Editor::get_output_format

adamsilverstein committed:
Improve tests for 6821

adamsilverstein committed:
Merge branch 'trunk' into ticket/56442

adamsilverstein committed:
test_quality_with_image_conversion_file_sizes: Verify that the selected editor actually supports WebP output

adamsilverstein committed:
Merge branch 'trunk' into ticket/56442

# Conflicts:
#	tests/phpunit/tests/media.php

adamsilverstein committed:
Remove empty line added inadvertantly

adamsilverstein committed:
Remove unrelaeted tests after merge

adamsilverstein commented on src/wp-includes/class-wp-image-editor.php:
```diff
@@ -414,8 +414,8 @@ protected function get_output_format( $filename = null, $mime_type = null ) {
 				$this->output_mime_type = $mime_type;
-				$this->set_quality();
 			}
+			$this->set_quality();
```
the is the only change: we always call set_quality here instead of only calling it at the initial transform. the remaining changes are test related.

adamsilverstein committed:
Merge branch 'trunk' into ticket/56442

felixarntz commented on tests/phpunit/tests/image/editor.php:
```diff
@@ -116,17 +116,26 @@ public function test_set_quality() {
 	 * @ticket 6821
 	 */
 	public function test_set_quality_with_image_conversion() {
+		remove_filter( 'image_editor_output_format', 'wp_default_image_output_mapping' );
```
This should be removed as it's now no longer in core.

adamsilverstein replied:
`image_editor_output_format` is still in core (https://github.com/WordPress/wordpress-develop/blob/d013b20c323c820e930351b89c93e8bcbccf8089/src/wp-includes/class-wp-image-editor.php#L378) it is applied to the output format since 5.8. I believe removing this is required to make this test pass but I will double check

adamsilverstein replied:
This change isn't needed, removing.

felixarntz commented on tests/phpunit/tests/media.php:
```diff
@@ -3621,6 +3621,105 @@ private function reset_omit_loading_attr_filter() {
+			$this->assertLessThan( $webp_sizes['sizes'][ $size ]['filesize'], $jpeg_sizes['sizes'][ $size ]['filesize'] );
+		}
+
+		// Cleanup.
+		remove_filter( 'wp_editor_set_quality', array( $this, 'image_editor_change_quality_low_jpeg' ), 10, 2 );
+		remove_filter( 'big_image_size_threshold', array( $this, 'add_big_image_size_threshold' ) );
```
Filters added in a test method don't need to be removed, as they get automatically removed. Also, this would be unreliable, as any failing assertion above would mean this code is never reached.

adamsilverstein replied:
removed here and in a couple of other tests that were removing added filters on the same hook

felixarntz reviewed (approved):
@adamsilverstein LGTM, except two minor nit-picks. Once they are removed, this should be clean and good to commit.

adamsilverstein committed:
Merge branch 'trunk' into ticket/56442

adamsilverstein committed:
Remove filter change in test

adamsilverstein committed:
Skip removing added filters at end of tests: all filters are removed automatically between test runs

peterwilsoncc commented on tests/phpunit/tests/media.php:
```diff
@@ -3621,6 +3621,101 @@ private function reset_omit_loading_attr_filter() {
+		}
+
+		$attachment_id = self::factory()->attachment->create_object(
+			$file,
+			0,
+			array(
+				'post_mime_type' => 'image/jpeg',
+			)
+		);
```
```suggestion
		$attachment_id = self::factory()->attachment->create_object(
			array(
				'post_mime_type' => 'image/jpeg',
				'file'           => $file,
			)
		);
```

You're using the deprecated argument format but if possible it would be good to use the `self::$large_id` shared fixture if possible. 

My understanding is that edits use new file names so it shouldn't be destructive but correct me if I am wrong.

adamsilverstein replied:
Ah, I didn't realize that, I copied the call from a nearby test. I'll commit your suggestion and verify tests still pass.

peterwilsoncc commented on tests/phpunit/tests/media.php:
```diff
@@ -3621,6 +3621,101 @@ private function reset_omit_loading_attr_filter() {
+		// be larger than the the matching generated JPEGs.
+		add_filter( 'wp_editor_set_quality', array( $this, 'image_editor_change_quality_low_jpeg' ), 10, 2 );
+
+		$editor = wp_get_image_editor( $file );
+
+		// Verify that the selected editor actually supports WebP output.
+		if ( ! $editor->supports_mime_type( 'image/webp' ) ) {
+			$this->markTestSkipped( 'WebP is not supported by the selected image editor.' );
+		}
```
Looks to double up the `wp_image_editor_supports` call above.

adamsilverstein replied:
Good point, removing

adamsilverstein replied:
removed

peterwilsoncc commented on tests/phpunit/tests/media.php:
```diff
@@ -3621,6 +3621,101 @@ private function reset_omit_loading_attr_filter() {
+		remove_filter( 'image_editor_output_format', array( $this, 'image_editor_output_webp' ) );
+
+		// The main (scaled) image: the JPEG should be smaller than the WebP.
+		$this->assertLessThan( $webp_sizes['filesize'], $jpeg_sizes['filesize'] );
```
```suggestion
		$this->assertLessThan( $webp_sizes['filesize'], $jpeg_sizes['filesize'], 'The JPEG should be smaller than the WebP.' );
```

:1234: As this test has multiple assertions, they'll need custom messages for each. For the assertion below, it would be good to include `$size` in the message so it can be determined which iteration fails.

peterwilsoncc commented on tests/phpunit/tests/image/editor.php:
```diff
@@ -119,14 +116,21 @@ public function test_set_quality_with_image_conversion() {
-		// Quality should change to the output format's value.
-		// A PNG image will be converted to WEBP whose quialty should be 86.
+		// When saving, quality should change to the output format's value.
+		// A PNG image will be converted to WEBP whose quality should be 86.
```
Nitty nit nit: Here it is `WEBP` but in the other file it is `WebP`

adamsilverstein replied:
`WebP` is correct, will fix.

adamsilverstein replied:
updated all uses here

peterwilsoncc reviewed (commented):
I've added a few minor notes within the tests. Still manually testing the fix though.

adamsilverstein committed:
Update tests/phpunit/tests/media.php

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

adamsilverstein committed:
Correct capitalization of WebP

adamsilverstein committed:
Merge branch 'ticket/56442' of github.com:adamsilverstein/wordpress-develop-fork into ticket/56442

adamsilverstein committed:
Remove duplicate call to wp_image_editor_supports in test

adamsilverstein committed:
Update tests/phpunit/tests/media.php

Co-authored-by: Peter Wilson <519727+peterwilsoncc@users.noreply.github.com>

adamsilverstein commented:
@peterwilsoncc thanks for the review and feedback. I addressed all your points and assuming tests pass this should be ready for commit.

>  Still manually testing the fix though.

A good way to test this filter is to output WebPs and try setting the quality very high or low, then checking your output files. A filter based on mime type won't work correctly before this fix. This is what the new test does to validate the fix, if you remove the one line change in `src/wp-includes/class-wp-image-editor.php` the new test will fail.

Here are the filters to use:

```php
// Output WebP.
add_filter( 'image_editor_output_format', function() {
	return array(
		'image/jpeg' => 'image/webp',
	);
});
```

```php 
// Filter WebP quality.
add_filter( 'wp_editor_set_quality', function( $quality, $mime_type ) {
	if ( 'image/webp' === $mime_type ) {
		return 99; // Very high quality, files will be very much larger than usual. Alternately, try a very low value.
	}
	return $quality;
}, 10, 2 );
```

adamsilverstein committed:
Revert "Remove duplicate call to wp_image_editor_supports in test"

This reverts commit a1f238a227082e94342612e6f4e5e7f6dd3ad5d4.

adamsilverstein committed:
Merge branch 'ticket/56442' of github.com:adamsilverstein/wordpress-develop-fork into ticket/56442

adamsilverstein committed:
Merge branch 'trunk' into ticket/56442

adamsilverstein committed:
Remove initial WebP support check to avoid duplication

adamsilverstein committed:
Merge branch 'trunk' into ticket/56442
